posts/Rust-1.86.0.md

```diff
@@ -0,0 +1,163 @@
+
+Trait upcasting may be especially useful with the `Any` trait, as it allows upcasting your trait object to `dyn Any` to call the downcast methods. Whereas before you'd have to write workarounds for this or use external crates.
+
+```rust
+use std::any::Any;
+
+trait MyAny: Any {}
+
+impl dyn MyAny {
+    fn downcast_ref<T>(&self) -> Option<&T> {
+        (self as &dyn Any).downcast_ref()
+    }
+}
+```
+
+You can [learn more about trait upcasting in the Rust reference](https://doc.rust-lang.org/reference/type-coercions.html#unsized-coercions).
+
+### `HashMap`s and slices now support indexing multiple elements mutably
+
+The borrow checker will prevent simaltaneous usage of references returned from `get_mut` methods. To get around this the standard library now provides a `get_disjoint_mut` helper on slices and `HashMap` to retrieve mutable references to multiple elements simaltaneously. See the following example taken from the API docs of [`slice::get_disjoint_mut`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.get_disjoint_mut):
+```rust
+let v = &mut [1, 2, 3];
+if let Ok([a, b]) = v.get_disjoint_mut([0, 2]) {
+    *a = 413;
+    *b = 612;
+}
+assert_eq!(v, &[413, 2, 612]);
+
+if let Ok([a, b]) = v.get_disjoint_mut([0..1, 1..3]) {
+    a[0] = 8;
+    b[0] = 88;
+    b[1] = 888;
+}
+assert_eq!(v, &[8, 88, 888]);
+
+if let Ok([a, b]) = v.get_disjoint_mut([1..=2, 0..=0]) {
+    a[0] = 11;
+    a[1] = 111;
+    b[0] = 1;
+}
+assert_eq!(v, &[1, 11, 111]);
+```
+
+### Allow safe functions to be marked with the #[target_feature] attribute.
+
+Functions marked with `#[target_feature]` are generally considered as unsafe functions: they are unsafe to call, cannot *generally* be assigned to safe function pointers, and don't implement the `Fn*` traits.
+
+However, calling them from other `#[target_feature]` functions with a superset of features is safe.
+
+```rust
+// Demonstration function
+#[target_feature(enable = "avx2")]
+fn avx2() {}
+
+fn foo() {
+    // Calling `avx2` here is unsafe, as we must ensure
+    // that AVX is available first.
+    unsafe {
+        avx2();
+    }
+}
+
+#[target_feature(enable = "avx2")]
+fn bar() {
+    // Calling `avx2` here is safe.
+    avx2();
+}
+```
+
+### Debug assert pointers are non-null when required for soundness
+
+The compiler will now insert checks that a pointer is not null upon non-zero-sized reads and writes, and also when the pointer is reborrowed into a reference. For example, the following code will now produce a non-unwinding panic:
+```rust
+let _x = *std::ptr::null::<u8>();
+let _x = &*std::ptr::null::<u8>();
+```
+Trivial examples like this have produced a warning since Rust 1.53.0, the new runtime check will detect these scenarios regardless of complexity.
```

Comment:
```suggestion
Trivial examples like this have produced a warning since Rust 1.53.0. The new runtime check will detect these scenarios regardless of complexity.
```